app/models/application_record.rb

```diff
@@ -20,6 +20,7 @@ def image_variant_url_for(variant)
     if ENV["S3_BUCKET"].present? && variant.service.public?
       variant.processed.url
     else
+      variant.blob.save! unless variant.blob.persisted?
```

Comment:
What if we don't provide a URL for unpersisted blobs? I'm just wondering in which cases the blob isn't persisted. It doesn't seem intuitive that a _url_for_ helper changes the database.

Reply:
Thanks for the early feedback @mkllnk. I had some cycles to work on this again and opened [the PR for the small refactor](https://github.com/openfoodfoundation/openfoodnetwork/pull/13342). 

> What if we don't provide a URL for unpersisted blobs? I'm just wondering in which cases the blob isn't persisted. It doesn't seem intuitive that a url_for helper changes the database.

You're absolutely right, it's not ideal for `url_for` to persist a blob. This was a quick fix to get CI passing after running into Rails 7.1's stricter checks. I'll revisit this in a future session and explore a cleaner approach, likely skipping URL generation for unpersisted blobs as you suggested. Thanks again for catching it!

Reply:
I followed your suggestion @mkllnk  and added an explicit exception for this edge case to keep the behavior consistent with expectations in `images_spec`: https://github.com/openfoodfoundation/openfoodnetwork/pull/13403.